src/shaka_scheme/system/vm/compiler/Compiler.cpp

```diff
@@ -11,34 +11,135 @@ Compiler::Compiler() {}
 
 Compiler::~Compiler() {}
 
-Expression Compiler::compile(Expression input, Expression next_instruction) {
+Expression Compiler::compile(Expression input, Expression
+next_instruction) {
   // (symbol? input)
-  // (list 'refer x next)
   if (is_symbol(input)) {
     Symbol instruction("refer");
     Data instruction_data(instruction);
 
-    DataPair next(*input, *list(create_node(*next_instruction)));
-    DataPair assembly(instruction_data, next);
-
-    return create_node(assembly);
+    return list(create_node(instruction_data), input, next_instruction);
   }
   // (pair? input)
   else if (is_pair(input)) {
-    // (list 'constant obj next)
-    if(car(input)->get<Symbol>() == Symbol("quote")){
+
+    Symbol& expression_type = car(input)->get<Symbol>();
+
+    // quote case
+    if (expression_type == Symbol("quote")) {
       Symbol instruction("constant");
       Data instruction_data(instruction);
 
-      DataPair next(*cdr(input), *list(create_node(*next_instruction)));
-      DataPair assembly(instruction_data, next);
+      return list(create_node(instruction_data), cdr(input), next_instruction);
```

Comment:
This seems dubious to me. If input is the list `(quote a)`, if I take the cdr of that, I should get back the list `(b)`, not the symbol 'b. However, I looked at your test case for compilation of `quote` and it appears to be working correctly, so I think this has to do with the questionable way in which the `cdr()` method is handling proper lists or pairs that represent proper lists. No biggie as long as the output is correct, but just want you to be aware that optically this looks wrong from a scheme semantics perspective.

Reply:
@Mortrax So in this case, I should be calling `car(cdr(input))`?

Reply:
@dylannakahodo That is correct. If it is indeed the list `(quote <obj>)`, `obj` is at the `(car (cdr ..))` position. That should be the only change needed.